File: julia/src/quantum_state.jl
"""
Immutable, validated quantum state container.

Three canonical representations are supported:
- `:vector`                   \u2014 state vector of shape (2^N,)
- `:density_matrix`           \u2014 density matrix of shape (2^N, 2^N)
- `:vectorized_density_matrix`\u2014 row-stacked vec(\u03c1) of shape (4^N,)

The non-canonical forms are computed on demand.  Mixed states (purity < 1)
are fully supported when constructed from a density matrix or its
vectorised form; accessing `state_vector` on a mixed state throws.
"""
module QuantumStates

export QuantumState,
       from_vector,
       from_density_matrix,
       from_vectorized_density_matrix,
       state_vector,
       density_matrix,
       purity

using LinearAlgebra


# ============================================================================
# Type definition
# ============================================================================

"""
    QuantumState

Immutable validated quantum state.

# Fields
- `n_qubits::Int`
- `state::Union{Vector{ComplexF64}, Matrix{ComplexF64}}`
- `state_type::Symbol`  \u2014 one of `:vector`, `:density_matrix`, `:vectorized_density_matrix`
"""
struct QuantumState
    n_qubits  ::Int
    state     ::Union{Vector{ComplexF64}, Matrix{ComplexF64}}
    state_type::Symbol
end


# ============================================================================
# Private helpers
# ============================================================================

function _check_power_of_two(dim::Int)::Int
    n = round(Int, log2(dim))
    2^n != dim && throw(ArgumentError("Dimension $(dim) is not a power of 2."))
    return n
end

function _check_normalised(norm::Number, atol::Float64, label::String)
    if abs(real(norm) - 1.0) > atol || abs(imag(norm)) > atol
        throw(ArgumentError(
            "State is not normalised: $(label) norm = $(norm), expected 1.0."
        ))
    end
end


# ============================================================================
# Factory constructors
# ============================================================================

"""
    from_vector(psi::AbstractVector; atol::Float64 = 1e-14) -> QuantumState

Construct a `QuantumState` from a state vector. Validates that the dimension
is a power of 2 and that the vector is normalised.

# Throws
- `ArgumentError` if dimension is not a power of 2.
- `ArgumentError` if the vector is not normalised.
"""
function from_vector(
    psi::AbstractVector;
    atol::Float64 = 1e-14
)::QuantumState
    \u03c8 = Vector{ComplexF64}(psi)
    n_qubits = _check_power_of_two(length(\u03c8))
    _check_normalised(dot(\u03c8, \u03c8), atol, "state vector")
    return QuantumState(n_qubits, \u03c8, :vector)
end


"""
    from_density_matrix(\u03c1::AbstractMatrix; atol::Float64 = 1e-14) -> QuantumState

Construct a `QuantumState` from a density matrix.
Validates shape (square, power-of-2 dimension), Hermiticity, and trace.
Both pure and mixed states are accepted.

# Throws
- `ArgumentError` if the matrix is not square or its dimension is not a power of 2.
- `ArgumentError` if the matrix is not Hermitian.
- `ArgumentError` if `Tr(\u03c1) \u2260 1`.
"""
function from_density_matrix(
    \u03c1::AbstractMatrix;
    atol::Float64 = 1e-14
)::QuantumState
    size(\u03c1, 1) != size(\u03c1, 2) && throw(ArgumentError(
        "Density matrix must be square, got shape $(size(\u03c1))."
    ))
    n_qubits = _check_power_of_two(size(\u03c1, 1))
    isapprox(\u03c1, \u03c1'; atol = atol) || throw(ArgumentError("Density matrix is not Hermitian."))
    _check_normalised(tr(\u03c1), atol, "density matrix trace")
    return QuantumState(n_qubits, Matrix{ComplexF64}(\u03c1), :density_matrix)
end


"""
    from_vectorized_density_matrix(\u03c1_vec::AbstractVector; atol::Float64 = 1e-14) -> QuantumState

Construct a `QuantumState` from a vectorised (row-stacked) density matrix.
Validates that the length is a perfect square with a power-of-2 dimension,
and that the reshaped matrix is Hermitian with trace 1.

# Throws
- `ArgumentError` on shape, Hermiticity, or trace violations.
"""
function from_vectorized_density_matrix(
    \u03c1_vec::AbstractVector;
    atol::Float64 = 1e-14
)::QuantumState
    dim = round(Int, sqrt(length(\u03c1_vec)))
    dim^2 != length(\u03c1_vec) && throw(ArgumentError(
        "Length of vectorised density matrix ($(length(\u03c1_vec))) is not a perfect square."
    ))
    n_qubits = _check_power_of_two(dim)
    \u03c1 = reshape(Vector{ComplexF64}(\u03c1_vec), dim, dim)
    isapprox(\u03c1, \u03c1'; atol = atol) || throw(ArgumentError("Density matrix is not Hermitian."))
    _check_normalised(tr(\u03c1), atol, "density matrix trace")
    return QuantumState(n_qubits, Vector{ComplexF64}(\u03c1_vec), :vectorized_density_matrix)
end


# ============================================================================
# Derived quantities
# ============================================================================

"""
    dim(s::QuantumState) -> Int

Hilbert space dimension 2^n.
"""
dim(s::QuantumState)::Int = 2^s.n_qubits


"""
    state_vector(s::QuantumState) -> Vector{ComplexF64}

Return the state vector. For density-matrix or vectorised forms the
state vector is extracted via eigendecomposition (pure states only).

# Throws
- `ArgumentError` if the state is mixed (purity < 1).
"""
function state_vector(s::QuantumState)::Vector{ComplexF64}
    s.state_type === :vector && return s.state::Vector{ComplexF64}

    d = dim(s)
    \u03c1 = s.state_type === :density_matrix ?
        s.state::Matrix{ComplexF64} :
        reshape(s.state::Vector{ComplexF64}, d, d)

    isapprox(real(tr(\u03c1 * \u03c1)), 1.0; atol = 1e-14) ||
        throw(ArgumentError("State is not pure; state vector is undefined for mixed states."))

    F = eigen(Hermitian(\u03c1))
    idx = argmax(real.(F.values))
    return Vector{ComplexF64}(F.vectors[:, idx])
end


"""
    density_matrix(s::QuantumState) -> Matrix{ComplexF64}

Return the density matrix. For vector-canonical states, computes \u03c1 = |\u03c8\u27e9\u27e8\u03c8|.
For vectorised states, reshapes the stored array.
"""
function density_matrix(s::QuantumState)::Matrix{ComplexF64}
    if s.state_type === :vector
        \u03c8 = s.state::Vector{ComplexF64}
        return \u03c8 * \u03c8'
    elseif s.state_type === :density_matrix
        return s.state::Matrix{ComplexF64}
    else
        d = dim(s)
        return reshape(s.state::Vector{ComplexF64}, d, d)
    end
end


"""
    purity(s::QuantumState) -> Float64

Purity `Tr(\u03c1\u00b2)`. Returns `1.0` immediately for vector-canonical states.
"""
function purity(s::QuantumState)::Float64
    s.state_type === :vector && return 1.0
    \u03c1 = density_matrix(s)
    return real(tr(\u03c1 * \u03c1))
end

end # module QuantumStates

File: python/src/domcsis_epic_tinker_box/quantum_state/quantum_state.py
from __future__ import annotations

import numpy as np
from dataclasses import dataclass


@dataclass(frozen=True)
class QuantumState:
    """
    Immutable, validated quantum state.
    
    Stores either the state vector or the density matrix, but not both. the other representation is computed on demand and cached. For pure states, the state vector is the canonical representation. The density matrix is computed as rho = |psi><psi| when requested.
    
    For mixed states, the density matrix is the canonical representation. The state vector is not defined for mixed states and will raise an error if accessed.


    Attributes:
        n_qubits: Number of qubits.
        state: State represetnation - state vector (2^N), density matrix (2^N x 2^N) or vectorized density matrix (4^N).
        state_type: "vector" if state is a state vector, "density_matrix" if state is a density matrix, "vectorized_density_matrix" if state is a vectorized density matrix.
    
    Example usage:
        psi = np.array([1/np.sqrt(2), 0, 0, 1/np.sqrt(2)], dtype=complex)  # Bell state |Φ+⟩
        state = QuantumState.from_vector(psi)
        print("State vector:", state.vector)
        print("Density matrix:\n", state.density_matrix)
        print("Purity:", state.purity)
        
    """
    
    n_qubits: int
    state: np.ndarray
    state_type: str  # "vector", "density_matrix", or "vectorized_density_matrix"

    # ------------------------------------------------------------------ #
    # Factory constructors                                               #
    # ------------------------------------------------------------------ #

    @classmethod
    def from_vector(cls, psi: np.ndarray, atol: float = 1e-14) -> QuantumState:
        """
        Construct from a state vector. Validates shape and normalisation.

        Args:
            psi:  1D array of shape (2^N).
            atol: Tolerance for normalisation check.

        Raises:
            ValueError: If dimension is not a power of 2.
            ValueError: If state is not normalised.
        """
        
        psi = np.asarray(psi, dtype=complex)
        if psi.ndim != 1:
            raise ValueError(
                f"State vector must be 1D, got shape {psi.shape}."
            )
            
        n_qubits = _check_power_of_two(psi.shape[0])
        
        _check_normalised(psi @ psi.conj(), atol, "state vector")
        return cls(n_qubits=n_qubits, state=psi, state_type="vector")

    @classmethod
    def from_density_matrix(cls, rho: np.ndarray, atol: float = 1e-14) -> QuantumState:
        """
        Construct from a density matrix. Validates shape,
        Hermiticity, and trace.

        Args:
            rho:  2D array of shape (2^N, 2^N).
            atol: Tolerance for all checks.

        Raises:
            ValueError: If shape is not square or dimension not a power of 2.
            ValueError: If rho is not Hermitian.
            ValueError: If Tr(rho) != 1.
        """
        rho = np.asarray(rho, dtype=complex)
        
        if rho.ndim != 2 or rho.shape[0] != rho.shape[1]:
            raise ValueError(
                f"Density matrix must be square 2D array, got shape {rho.shape}."
            )
            
        n_qubits = _check_power_of_two(rho.shape[0])
        
        if not np.allclose(rho, rho.conj().T, atol=atol):
            raise ValueError("Density matrix is not Hermitian.")
        
        _check_normalised(np.trace(rho), atol, "density matrix trace")
        
        
        return cls(n_qubits=n_qubits, 
                   state=rho, 
                   state_type="density_matrix",
                   )


    @classmethod
    def from_vectorized_density_matrix(cls, rho_vec: np.ndarray, atol: float = 1e-14) -> QuantumState:
        """
        Constructs from a vectorized density matrix. Validates shape, Hermiticity, and trace.
        
        Args:
            rho_vec: 1D array of shape (4^N,).
            atol: Tolerance for all checks.
            
        Raises:
            ValueError: If shape is not 1D or dimension not a power of 4.
            ValueError: If reshaped rho is not Hermitian.
            ValueError: If Tr(rho) != 1.        
        """
        
        rho_vec = np.asarray(rho_vec, dtype=complex)
        
        if rho_vec.ndim != 1:
            raise ValueError(
                f"Vectorized density matrix must be 1D array, got shape {rho_vec.shape}."
            )
            
        n_qubits = _check_power_of_two(int(np.sqrt(rho_vec.shape[0])))
        
        rho = rho_vec.reshape((2 ** n_qubits, 2 ** n_qubits))
        
        if not np.allclose(rho, rho.conj().T, atol=atol):
            raise ValueError("Density matrix is not Hermitian.")
        
        _check_normalised(np.trace(rho), atol, "density matrix trace")
        
        return cls(
            n_qubits=n_qubits, 
            state=rho_vec, 
            state_type="vectorized_density_matrix",
            )
        
    # ------------------------------------------------------------------ #
    # Derived properties                                                   #
    # ------------------------------------------------------------------ #

    @property
    def dim(self) -> int:
        """Hilbert space dimension 2^n."""
        return int(2 ** self.n_qubits)

    @property
    def density_matrix(self) -> np.ndarray:
        """Compute rho = |psi><psi| on demand."""
        return np.outer(self.vector, self.vector.conj())

    @property
    def vector(self) -> np.ndarray:
        """
        Returns the state vector from a density matrix or a vectorized density matrix if that is pure, else raises an error.
        """
        
        if self.state_type == "vector":
            return self.state
        
        if self.state_type == "density_matrix":
            rho = self.state
        elif self.state_type == "vectorized_density_matrix":
            rho = self.state.reshape((self.dim, self.dim))
        else:
            raise ValueError(f"Invalid state_type: {self.state_type}")
        
        # Check purity: Tr(rho^2) == 1 for pure states
        if not np.isclose(np.trace(rho @ rho), 1.0, atol=1e-14):
            raise ValueError("State is not pure, cannot extract state vector.")
        
        # For a pure state, the eigenvector with eigenvalue 1 is the state vector
        eigvals, eigvecs = np.linalg.eigh(rho)
        idx = np.argmax(eigvals)
        return eigvecs[:, idx]
    
    
    # =============================================================================
    # State manipulation methods (e.g. partial trace) could be added here in the future. 
    # For now, this class is just a validated container for quantum states. 
    # ============================================================================
    
    @property
    def purity(self) -> float:
        """Purity Tr(rho^2). 1 for pure states, <1 for mixed states."""
        if self.state_type == "vector":
            return 1.0
        elif self.state_type == "density_matrix":
            rho = self.state
        elif self.state_type == "vectorized_density_matrix":
            rho = self.state.reshape((self.dim, self.dim))
        else:
            raise ValueError(f"Invalid state_type: {self.state_type}")
        
        return float(np.real(np.trace(rho @ rho)))
    

# ============================================================================
# Private helpers
# ============================================================================

def _check_power_of_two(dim: int) -> int:
    """Return log2(dim) if dim is a power of 2, else raise."""
    n = int(np.log2(dim))
    if 2 ** n != dim:
        raise ValueError(
            f"State dimension {dim} is not a power of 2."
        )
    return n


def _check_normalised(norm: complex, atol: float, label: str) -> None:
    if not np.isclose(norm.real, 1.0, atol=atol) or abs(norm.imag) > atol:
        raise ValueError(
            f"State is not normalised: {label} norm = {norm:.7f}, expected 1.0."
        )